+ETC1 Atlas Compressor

JFixby opened the pull request:
Back-story: https://twitter.com/badlogicgames/status/643082568334745600

```AtlasETC1Compressor``` takes atlas produced by the ```TexturePacker``` and converts it to ```ETC1``` format by:
- compressing textures from ```PNG``` to ```ETC1```,
- changing texture-file names accordingly  in the atlas description file.

To test the ```AtlasETC1Compressor``` please run the ```AtlasETC1CompressorTest```, the test will:
- prepare example atlas using sprites in the sprites folder,
- compress the atlas using ```AtlasETC1Compressor```,
- load and render compressed sprites on the screen using gdx desktop backend.

intrigus commented:
Please, use camelCase and not underscore_case.

Tom-Ski commented:
Please use proper java naming conventions. https://github.com/libgdx/libgdx/wiki/Contributing
Would rather have the test outside of the tools package too.

intrigus commented on extensions/gdx-tools/src/com/badlogic/gdx/tools/etc1/AtlasETC1Compressor.java:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	private static void reportError (String err_message) {
+		throw new Error(err_message);
```
Why are you throwing an [Error](https://docs.oracle.com/javase/7/docs/api/java/lang/Error.html)?
I'd instead use a `RuntimException`.
I'd also remove this method, since it's unnecessary, why do you not directly throw an Exception?
No need for this helper method in my opinion :)

Darkyenus replied:
Or `NullPointerException` (with a message what was null), let's use appropriate exceptions.

JFixby replied:
Imagine you have a project with 800 different exceptions thrown by ```throw someException();```
Now you decide to deploy it on a server and you want all the errors to be reported (sent) to another (log) server. Now you have to go to all these 800 places and fix each of them by introducing the helper.

Usually I use in my code ```Error.reportError(...)``` where  the Error is a static class (similar to Gdx.log) containing current logger system that can vary from simple ```System.out.println()``` to an ```Apache Active MQ```-client. Since you don't do anything like this on libGDX I decided to wrap the error it in the simple method, allowing you to decide what to do with that, and leaving you a freedom to introduce any logic you prefer in the future.

intrigus commented on extensions/gdx-tools/src/com/badlogic/gdx/tools/etc1/AtlasETC1CompressionResult.java:
```diff
@@ -0,0 +1,59 @@
+
+		@Override
+		public String toString () {
+			return "" + old_page_file_name + " :-> " + new_page_file_name + "";
```
I'd replace this with 
````java
return old_page_file_name + " :-> " + new_page_file_name;
````

JFixby replied:
I can do it but java compiler will do it anyway.

MobiDevelop replied:
Please don't do silly things based on the assumption that "the compiler will make it not silly".

JFixby replied:
It was left in case you want to use square brackets to wrap the output.

MobiDevelop replied:
I thought it was left in because the compiler will fix it anyway? Anyway, please don't add extraneous things on the basis that someone might want to format it in a specific way in the future. Make the code clean and easy to read by default...  if someone wants to add more formatting, they will come along and do it when they choose.

Darkyenus commented on extensions/gdx-tools/src/com/badlogic/gdx/tools/etc1/AtlasETC1CompressorTest.java:
```diff
@@ -0,0 +1,135 @@
+
+	public static void main (String[] args) throws Exception {
+
+		String application_home_path_string = System.getProperty("user.dir");
```
Why does the test use user's home directory?

JFixby replied:
It's not user's home directory. This is a complete absolute path from where the application was initialized.

Darkyenus commented:
I like and appreciate what this does, the code however feels somewhat enterprise-y. It adds 4 classes just to call `ETC1Compressor.process(...)`. Wouldn't be better if there was just a settings flag for TexturePacker to output ETC1 instead of PNG?

_It could compress to ETC1 directly, without ever compressing and decompressing through PNG (it would just have to convert awt BufferedImage to Pixmap, if I'm reading the code right)._

JFixby commented:
> Would rather have the test outside of the tools package too.

@Tom-Ski 
I placed the test inside because the https://github.com/libgdx/libgdx/blob/master/extensions/gdx-tools/src/com/badlogic/gdx/tools/texturepacker/TexturePackerTest.java is also inside.
Do you want me to create a new project, like **gdx-tools-test** and move the tests there?

JFixby commented:
@Darkyenus Generally It is ok to create more classes as long as they are easily readable and maintainable. In the proposed implementation ```AtlasETC1Compressor``` is constructed on top of the TexturePacker avoiding any potential bugs and simplifying code maintenance for both of them.

Darkyenus commented:
But there is no reason why `newCompressionSettings()` should exist. There is nothing it can do which constructor can't. Furthermore, AFAIK no libGDX settings classes use this pattern. But that is just a nitpick, it still feels overengineered for what is basically one method call wrapper.

JFixby commented:
Ok, I made some updates:
- fixed the camelCase
- renamed AtlasETC1Compressor* to ETC1AtlasCompressor*
- git-ignored test output
- updated the test to render two atlases for visual comparison

Now I need input about **gdx-tools-test**. Maybe it makes sense to merge first and then someone can pull a new PR about the **gdx-tools-test** because I don't know hot it suppose to be done anyway.

JFixby commented:
Off-Topic:

@Darkyenus 

Here is the proper scenario for the pattern usage:
```
//before start
FancyObjectFactory.installComponent(new SomeFancyObjectFactoryImplementation());

//in action
FancyObjectSpecs specs = FancyObjectFactory().newSpecs();
    ...setup specs
FancyObject fancy_object = FancyObjectFactory().newObject(specs );
fancy_object .use();
```

```FancyObjectSpecs``` and ```FancyObject ``` must be java interfaces with no exceptions.

Now without changing the action code above we can perform the following scenarious:

Usage scenario A: local usage
```FancyObjectFactory.installComponent(new LocalFancyObjectFactory());```
```.newSpecs()``` and .```newObject(specs)``` - return regular classes ```SpecsImpl```and ```FancyObjectImpl```

Usage scenario B: remote usage
```FancyObjectFactory.installComponent(new RemoteFancyObjectFactory());```
```.newSpecs()``` and .```newObject(specs)``` - perform remote call to a remote server and return handlers instead of the actual objects. The server performs the scenario A.

Usage scenario C: testing and analysis
```FancyObjectFactory.installComponent(new SystemAnalytycsFancyObjectFactory());```

Usage scenario D: trying older version of the library
```FancyObjectFactory.installComponent(new OldVersionOfTheFancyObjectFactory());```

Usage scenario E: duplicate all calls to guarantee the stability of the production code
```FancyObjectFactory.installComponent(new BridgeFancyObjectFactory(new RemoteFancyObjectFactory(server with old library  version),new RemoteFancyObjectFactory(server with new library version)));```

...more...

Using simple constructor ```FancyObjectSpecs specs = new FancyObjectSpecs ()``` locks you in a single scenario A, because in the future replacing the ```FancyObjectSpecs``` class with an interface will break the api. While the ```newCompressionSettings()``` gives some flexibility on that.

intrigus commented on extensions/gdx-tools/src/com/badlogic/gdx/tools/etc1/ETC1AtlasCompressor.java:
```diff
@@ -0,0 +1,89 @@
+		return null;
+	}
+
+	private static void reportError (String errMessage) {
```
This method is still unnecessary. Nobody needs it imo :)

davebaol replied:
Couldn't agree more.

JFixby replied:
@davebaol https://github.com/libgdx/libgdx/pull/3907#discussion_r55281908

davebaol replied:
TBH I don't think that this is a proper solution to the problem you mentioned, but still...

davebaol commented on extensions/gdx-tools/src/com/badlogic/gdx/tools/etc1/ETC1AtlasCompressionResult.java:
```diff
@@ -0,0 +1,59 @@
+
```
Missing license header.
The same goes for the other classes too.

JFixby replied:
How about the http://unlicense.org/ ?
```
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
```

JFixby replied:
Header: 
```
/*******************************************************************************
 * Licensed under the Unlicense license (the "License");
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this file, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 * You may obtain a copy of the License at
 * 
 *   http://unlicense.org
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.

 ******************************************************************************/
```

davebaol replied:
Uh? Why not just Apache License 2.0 like everything else in libgdx?

JFixby replied:
I'll take it as a green light.

JFixby replied:
Practically the differences between all open source licenses exist only on paper, while in fact they all are the Unlicense. Anyway the Unlicense allows anyone to use it under the Apache too.

MobiDevelop replied:
Are you purposely trying to be difficult?

JFixby replied:
It is difficult to build a meaningful dialog here since there is no specific person assigned to evaluate this PR.

MobiDevelop replied:
This is open source software, any contributor can/may review it. Purposely choosing a different license than that of the rest of of the project is being difficult. Not following the contributing guidelines is being difficult. I don't think this is a side-effect of not having an assigned person to review the PR.

JFixby replied:
I'll replace the license to make it aligned with the rest of of the project.

Darkyenus commented:
@JFixby I agree that this can be an useful pattern - for enterprise software. This code is literally just a boilerplate for calling _one_ method of `ETC1Compressor` and doing string replacing. It is very unlikely that we'll ever want to add more setting parameters than just that one String path. And if we do, we can just add one more method with another parameter (as libGDX already does elsewhere) or full settings class, if there are more parameters.

Furthermore, it can't be used from the command line, like `TexturePacker`. So if I were to use it in my toolchain, I could probably get this functionality in about the same amount of code without using this PR, so I am not quite sure what problem is this even solving.

JFixby commented:
@Darkyenus Well I guess I can make it command line-oriented.

JFixby commented:
Seems like test-server unable to download the http://libgdx.badlogicgames.com/nightlies/dist//gdx-natives.jar
How to rerun the test?

JFixby commented:
I'll re-submit it to make the commits more clear: https://github.com/libgdx/libgdx/pull/3909
